pallets/messages/src/lib.rs

```diff
@@ -0,0 +1,268 @@
+		type MaxMessageSizeInBytes: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_block_messages)]
+	pub(super) type BlockMessages<T: Config> //TODO: convert Vec to BoundedVec
+		= StorageValue<_, Vec<(Message<T::AccountId>, SchemaId)>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_messages)]
+	pub(super) type Messages<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		T::BlockNumber,
+		Twox64Concat,
+		SchemaId,
+		Vec<Message<T::AccountId>>,
+		ValueQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Too many messages are added to existing block
+		TooManyMessagesInBlock,
+		/// Message size is too large
+		TooLargeMessage,
+		/// Invalid Pagination Request
+		InvalidPaginationRequest,
+		/// Type Conversion Overflow
+		TypeConversionOverflow,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Messages are stored for a specified schema id and block number
+		MessagesStored { schema_id: SchemaId, block_number: T::BlockNumber, count: u16 },
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(current: T::BlockNumber) -> Weight {
+			let prev_block = current - T::BlockNumber::one();
+			Self::move_messages_into_final_storage(prev_block)
+			// TODO: add retention policy execution
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Adds a message into storage
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schema_id`: Registered schema id for current message
+		/// - `message`: Serialized message data
+		///
+		/// Result is equivalent to the dispatched result.
+		///
+		/// # <weight>
+		/// Execution weight
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::add(message.len() as u32, 1_000))]
+		pub fn add(origin: OriginFor<T>, schema_id: SchemaId, message: Vec<u8>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				message.len() < T::MaxMessageSizeInBytes::get().try_into().unwrap(),
+				Error::<T>::TooLargeMessage
+			);
+			// TODO: validate schema existence and validity from schema pallet
+			// TODO: get MsaId from mrc-identity pallet
+
+			let mut block_messages = <BlockMessages<T>>::get();
+			let current_size: u16 = block_messages
+				.len()
+				.try_into()
+				.map_err(|_| Error::<T>::TypeConversionOverflow)?;
+			ensure!(
+				current_size < T::MaxMessagesPerBlock::get(),
+				Error::<T>::TooManyMessagesInBlock
+			);
+
+			let m = Message {
+				data: message,
+				signer: who,
+				index: current_size,
+				msa_id: 0 as MessageSenderId, // TODO: replace with actual msa_id
+			};
+			block_messages.push((m, schema_id));
+			<BlockMessages<T>>::set(block_messages);
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn get_messages_by_schema(
```

Comment:
This is not being used. Please remove from this PR.

Reply:
Since you are feeling really strongly about this I'll remove it.